### PR TITLE
 Configurable ingress port

### LINKED ISSUE #213: ingress gateway tls port should be customisable
At present, it is only possible for the ingress gateway TLS port to listen on 443.
It should be possible to customise the TLS port - e.g. configure it to listen on `8443`.
This needs to be an operator configuration - and the `ingress_controller.go` needs to be modified to remove the
hardcoded `api.Spec.ListenPort = 443`, preferring to reference the operators startup configuration:

```
				for _, tls := range desired.Spec.TLS {
					for _, host := range tls.Hosts {
						if hostName == host {
							api.Spec.Protocol = "https"
							api.Spec.CertificateSecretNames = []string{
								tls.SecretName,
							}
							api.Spec.ListenPort = 443
						}
					}
				}
```

### LINKED ISSUE #284: TLS ingress should not try to open 443 on the Tyk container
## Expected Behavior

443 is usually not able to be opened on our containers with default perms

## Current Behavior

If an API is templated by the ingress controller it sets the API port to 443. 

## Possible Solution

Make this configurable by the user and set a sensible default such as 8443. We then do the mapping at the ingress/service level to the container port

## Steps to Reproduce

Try and run the TLS ingress examples

## Context
This is a blocker for several clients.

### PR DESCRIPTION
**Pull request by gernest**
fixes #284
fixes #213 

# env vars

- `TYK_HTTPS_INGRESS_PORT`
- `TYK_HTTP_INGRESS_PORT`

# helm

```
ingressHTTPSPort: 8443
ingressHTTPPort: 8080
```

### CODE REVIEW

File: config/helm/manager_env_vars_patch.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  name: controller-manager
+  namespace: system
+spec:
+  template:
+    spec:
+      containers:
+        - name: manager
+          env:
+            - name: TYK_TLS_INGRESS_PORT
+              value: "433"
```
Comment (asoorm) [inline config/helm/manager_env_vars_patch.yaml:13]: Can we expose:

```
TYK_HTTP_INGRESS_PORT
```

AND

```
TYK_HTTPS_INGRESS_PORT
```

File: docs/installation/installation.md [hunk 2]
```diff
@@ -54,6 +54,7 @@ Operator configurations are all stored in the secret `tyk-operator-conf`.
 | `TYK_TLS_INSECURE_SKIP_VERIFY` | `true`                                     | If the Tyk URL is HTTPS and has a self-signed certificate; defaults to `false`                             |
 | `WATCH_NAMESPACE`              | `foo,bar`                                  | Comma separated list of namespaces for Operator to operate on; defaults to all namespaces if not specified |
 | `WATCH_INGRESS_CLASS`          | `customclass`                              | Default `tyk` if omitted; allows Tyk Operator to watch a different ingress class                           |
+| `TYK_TLS_INGRESS_PORT`          | `443`                              | Default `8443` if omitted; sets ListenPort for TLS ingress
```
Comment (asoorm) [inline docs/installation/installation.md:57]: should be

```
| `TYK_TLS_INGRESS_PORT` | `8443` | Default `8443` if omitted; sets ListenPort for HTTPS ingress
```

File: pkg/environmet/env.go [hunk 3]
```diff
@@ -76,6 +82,11 @@ func (e *Env) Parse() error {
 	e.Auth = strings.TrimSpace(os.Getenv(TykAuth))
 	e.Org = strings.TrimSpace(os.Getenv(TykORG))
 	e.InsecureSkipVerify, _ = strconv.ParseBool(os.Getenv(SkipVerify))
+	e.IngressTLSPort, _ = strconv.Atoi(os.Getenv(IngressTLSPort))
+	if e.IngressTLSPort == 0 {
+		// set default value
+		e.IngressTLSPort = 8442
```
Comment (asoorm) [inline pkg/environmet/env.go:88]: 8442?

File: pkg/environmet/env.go [hunk 4]
```diff
@@ -65,6 +71,15 @@ func (e Env) Merge(n Env) Env {
 	if n.IngressClass != "" {
 		e.IngressClass = n.IngressClass
 	}
+	if n.IngressTLSPort != 0 {
+		e.IngressTLSPort = n.IngressTLSPort
+	}
+	if n.IngressTLSPort != 0 {
+		e.IngressTLSPort = n.IngressTLSPort
+	}
```
Comment (asoorm) [inline pkg/environmet/env.go:79]: This looks like duplicate

File: hack/helm/pre_helm.go [hunk 5]
```diff
@@ -49,6 +51,13 @@ const annotationsDest = `  template:
       labels:
         control-plane: controller-manager`
 
+const envHTTPSSrc = `value: "433"`
+
+const envHTTPSDEST = `value: {{default 8443 .Values.ingressHTTPSPort |quote}}`
+const envHTTPSrc = `value: "8080"`
+
+const envHTTPDEST = `value: {{.Values.ingressHTTPPort |quote}}`
```
Comment (asoorm) [inline hack/helm/pre_helm.go:59]: nit: maybe we can clean this up a little?

```
const (
        envHTTPSSrc = `value: "433"`
        envHTTPSDEST = `value: {{ default 8443 .Values.ingressHTTPSPort | quote }}`
        envHTTPSrc = `value: "8080"`
        envHTTPDEST = `value: {{ .Values.ingressHTTPPort | quote }}`
)
```

Comment (gernest) [reply hack/helm/pre_helm.go]: done 8b0052a

File: pkg/environmet/env.go (hunk 4 above)
Comment (gernest) [reply pkg/environmet/env.go]: nice catch! done 61c0bbf

File: ci/helm_values.yaml [hunk 6]
```diff
@@ -28,7 +30,8 @@ serviceAccount:
 annotations: {}
 podAnnotations: {}
 podSecurityContext: {}
-resources: {}
+resources:
```
Comment (asoorm) [inline ci/helm_values.yaml:33]: nit: this could go on same line `resources: {}`

File: config/helm/manager_env_vars_patch.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  name: controller-manager
+  namespace: system
+spec:
+  template:
+    spec:
+      containers:
+        - name: manager
+          env:
+            - name: TYK_HTTPS_INGRESS_PORT
+              value: "433"
```
Comment (asoorm) [inline config/helm/manager_env_vars_patch.yaml:13]: should this be `8443` by default?

File: hack/helm/pre_helm.go [hunk 8]
```diff
@@ -49,6 +51,13 @@ const annotationsDest = `  template:
       labels:
         control-plane: controller-manager`
 
+const (
+	envHTTPSSrc  = `value: "433"`
```
Comment (asoorm) [inline hack/helm/pre_helm.go:55]: `443` -> `8443`

**Review by asoorm — CHANGES_REQUESTED** [review]
few small changes

File: hack/helm/pre_helm.go (hunk 8 above)
Comment (gernest) [reply hack/helm/pre_helm.go:55]: this is what is on kustomize, we replace this with a helm template and `8443` as default.

Comment (asoorm) [reply hack/helm/pre_helm.go:55]: so let's set it to `8443 from the start right?`

File: config/helm/manager_env_vars_patch.yaml (hunk 7 above)
Comment (gernest) [reply config/helm/manager_env_vars_patch.yaml]: done a8185c9

File: hack/helm/pre_helm.go (hunk 8 above)
Comment (gernest) [reply hack/helm/pre_helm.go]: you are right, fixed a8185c9

**Comment by gernest** [comment]
@asoorm PTAL

File: ci/helm_values.yaml (hunk 6 above)
Comment (asoorm) [reply ci/helm_values.yaml:33]: Just this one

**Review by asoorm — APPROVED** [review]
Just the nit mentioned above. Apart from that it looks great. Thanks

**Pull request merged**